pkg/minikube/tunnel/tunnel_test.go

```diff
@@ -8,42 +8,82 @@ import (
 
 	"github.com/docker/machine/libmachine/state"
 	core_v1 "k8s.io/api/core/v1"
-	"k8s.io/minikube/pkg/util"
 )
 
 type StateTransitionTestCase struct {
 	description    string
-	observedState  *LoadBalancerTunnelState
-	expectedUpdate *LoadBalancerTunnelStateUpdate
+	observedState  *LBTunnelState
+	expectedUpdate *LBTunnelUpdate
 }
 
 func TestNoServicesNoRoutes(t *testing.T) {
 	test(StateTransitionTestCase{
 		description: "no services,no routes -> no change",
-		observedState: &LoadBalancerTunnelState{
+		observedState: &LBTunnelState{
 			ClusterState:             state.Running,
 			Routes:                   []Route{},
 			ServicesWithLoadBalancer: []core_v1.Service{},
 		},
-		expectedUpdate: &LoadBalancerTunnelStateUpdate{
+		expectedUpdate: &LBTunnelUpdate{
 			RoutesToDelete:   nil,
 			RoutesToAdd:      nil,
 			ServicesToUpdate: nil,
 		},
 	}, t)
 }
 
+
+func TestCleanupRoutesWhenNoServices(t *testing.T) {
+	test(StateTransitionTestCase{
+		description: "no services,some routes -> cleanup routes",
+		observedState: &LBTunnelState{
+			ClusterState:             state.Running,
+			Routes:                   someRoutes(),
+			ServicesWithLoadBalancer: []core_v1.Service{},
+		},
+		expectedUpdate: &LBTunnelUpdate{
+			RoutesToDelete:   someRoutes(),
+			RoutesToAdd:      nil,
+			ServicesToUpdate: nil,
+		},
+	}, t)
+}
+
+
+func TestStoppedMinikube(t *testing.T) {
+	for _, state := range []state.State{state.None, state.Error, state.Stopped} {
+		test(StateTransitionTestCase{
+			description: fmt.Sprintf("some routes + minikube state: [%s] -> delete all routes", state),
+			observedState: &LBTunnelState{
+				ClusterState:             state,
+				Routes:                   someRoutes(),
+				ServicesWithLoadBalancer: []core_v1.Service{},
+			},
+			expectedUpdate: &LBTunnelUpdate{
+				RoutesToDelete:   someRoutes(),
+				RoutesToAdd:      nil,
+				ServicesToUpdate: nil,
+			},
+		}, t)
+	}
+}
+
+func someRoutes() []Route {
+	routes := []Route{
+		{
+			IPRange: &net.IPNet{
```

Comment:
It makes sense to support ranges but in the current implementation of clusterip->externalip, I believe a route can be encoded as only an IP address w/ no mask

Reply:
well, yes, for those that are created by minikube tunnel. However as "read" routes with `netstat -nr` for example, there can be ranges there which were not created by us. I was thinking about checking for overlapping CIDRs too potentially?